desc/objectives/_coils.py

```diff
@@ -103,11 +96,12 @@ def build(self, use_jit=True, verbose=1):  # noqa:C901
 
         """
         # local import to avoid circular import
-        from desc.coils import CoilSet, MixedCoilSet, _Coil
+        from desc.coils import CoilSet, MixedCoilSet
 
         self._dim_f = 0
         self._quad_weights = jnp.array([])
 
+        # FIXME: Is this necessary? It only gets called on a Grid, never coils/coilset
```

Comment:
@daniel-dudt Yeah I'm pretty certain it's necessary. the input is a pytree whose leaves are Grid -- it is in the same pytree structure as the user-inputted coilset so that it can go through each grid to get the total dim_f

Reply:
If it is a pytree of Grids, then won't only the first and last if statements ever trigger? I don't see when the `CoilSet` checks will ever trigger. 
```
if isinstance(coilset, list):
    # this will trigger at the branches
elif isinstance(coilset, MixedCoilSet):
    # I don't think this is necessary
elif isinstance(coilset, CoilSet):
    # I don't think this is necessary
elif isinstance(coilset, _Grid):
    # this will trigger at the leaves
```

I think this could be simplified with something like the following: 
```
grids = tree_leaves(tree_of_grids, is_leaf=lambda x: hasattr(x, "num_nodes"))
self._dim_f = np.sum([grid.num_nodes for grid in grids])
self._quad_weights = np.concatenate([grid.spacing[:, 2] for grid in grids])
```

Reply:
It's a pytree of objects that are of type `CoilSet` or `MixedCoilSet` but have Grids as leaves so every elif statement is used. I think Dario had the exact same question a while back: https://github.com/PlasmaControl/DESC/pull/853#discussion_r1526764746

And using `hasattr(x, "num_nodes")` is smart, I didn't think to use that at the time. I'll have to try it but it seems like it would work

Reply:
thank you! :)